### DISCUSSION TITLE
How to speed up searching reference symbol? (textDocument/reference)

### DISCUSSION THREAD

**Discussion by amaksymov** [microsoft/pyright#3554] [Q&A]
I am working on a large code base and when I want to find all references to a function/method/class/variable and it takes a long time every time. How can I speed it up, maybe cache it or something like that?

As I understand it, pylance is responsible for this in VSCode, but pylance only comes with VSCode, in my case I use NeoVim. Maybe there are tools that speed up the search for links, but after a week of research I found nothing how to speed up that proccess.

I will be glad to any suggestions

**Comment by erictraut** [comment]
Pyright is focused on type checking, not on language server features like "find references". It does have a minimal reference implementation for some language server features, but there are limitations like the one you've noted here.

Pylance provides premium language server features including indexing that speeds up functionality like "find references". If you want these premium language server features, you would need to switch to VS Code. There are [extensions for VS Code](https://marketplace.visualstudio.com/items?itemName=vscodevim.vim) that provide Vim key bindings and other Vim functionality.

**Reply by amaksymov** [reply]
But how do [coc.nvim](https://github.com/neoclide/coc.nvim) with [coc-pyright](https://github.com/fannheyward/coc-pyright) achieve the same speed in the NeoVim? Do they use only pyright or 3rd party libs? Or do they implement their own logic, such as caching?
